rplugin/python3/deoplete/sources/ale.py

```diff
@@ -24,10 +24,13 @@ def __init__(self, vim):
         self.rank = 1000
         self.is_bytepos = True
         self.min_pattern_length = 1
+        self.input_pattern = r'(\.|::|->)\w*$'
```

Comment:
Is there some way we can supply different input patterns for different languages for Deoplete? `::` is valid for C++ or Rust, but not for JavaScript, for example.

One thing I want to eventually do is read the trigger characters from the language server responses, and make the ALE codebase totally language agnostic. It would be nice if it's possible to update the input pattern here based on language server messages in the future. You can see what I mean if you look at the LSP spec, in the `CompletionOptions` interface. https://microsoft.github.io/language-server-protocol/specification

Reply:
For possible solution with current implementation of Deoplete, I can use `input_patterns` option for this. This option can set `input_pattern`'s for several filetypes beforehand.

```python
self.input_patterns = {
  'javascript': r'\.\w*$',
  'python': r'\.\w*$',
  'c': r'(::|->)\w*$',  # is this true? I'm not familiar with c/c++ ;(
  ...
  ...
  '_': r'(\.|::|->)\w*$',  # This is fallback for other languages.
}
```

Each typical Deoplete source is for one filetype. But ALE source is different, that is, it can deal with ALL filetypes. So `input_patterns` should contain as many filetypes as possible.

I will commit this for PR, but review it and fill up lacked languages.

----

And one more about LSP spec. Surely, it can change `input_pattern` with response about `CompleteOptions`. But I wonder if it can be done as expected.

Because I think it make no sense to change `input_pattern` after invoking Deoplete. I think the option should be placed on each source itself, or called before invoking Deoplete. Can you tell your opnion, @Shougo ?

Reply:
From [this thread in Twitter (Japanese)][twitter], this implementation below is suggested by @Shougo.

[twitter]: https://twitter.com/delphinus35/status/1141529440851070976

1. At first, `input_pattern` should include all types of indicators that mean starting completion.

   ```python
   # Is this sufficient for all languages?
   self.input_pattern = r'(\.|::|->)\w*$'
   ```

2. When users invoke Deoplete and input texts end with indicators above, it should call LSP and fetch characters to trigger completion.
   - Is this `triggerCharacters` in LSP spec? (TODO for me: read the spec in detail)

3. If it matches one of the characters, `gather_candidates()` will return candidates. If not, it returns `[]`.

Reply:
For now, set up `input_pattern` with a `dict` like you suggested here. Have a look at the patterns used in `completion.vim` and try to match the different filetypes and the default pattern there. Leave a comment in both places indicating that people who update one part of the code should update the other part of the code, for people who add support for more languages later.

In the long term, it would be nice to set up these patterns automatically. The `triggerCharacters` attribute is part of the spec and used by many servers. What it means is that at some point in the future, it will be possible to set up the patterns for all languages in a totally language agnostic way, without having to come back and add support for more languages in these two files. The servers will tell ALE which characters completion should be done on. Don't worry about that right now.

Reply:
@w0rp I added definition for `input_patterns` and comments for both there and `completion.vim`. Check it.

I use `.` only for the fallback to triggers to match both definition. `(.|::|->)` is better for it, maybe.